### DISCUSSION TITLE
How to match the rendered image to the canvas size?

### DISCUSSION THREAD

**Discussion by Scooter-F** [udevbe/greenfield#138] [Q&A]
I've got an issue where the size of the application rendered through greenfield does not match the size of the canvas element it's being rendered in. The application is getting cut off. I had assumed that the size of the rendered application would match what's set for the canvas size, but is that assumption wrong? Is there another step I need to take? Any help would be super appreciated!

**Comment by Zubnix** [comment]
On initial render Greenfield will allow the application to choose it's own size. It seems this has an obvious downside 😅 I'll see if I can make a fix for this.

**Reply by Scooter-F** [reply]
That would be awesome, thanks! It's worth noting that the application I'm attempting to render is full screen by default. Not sure if that changes things.

**Reply by Zubnix** [reply]
This is now (Hopefully) fixed by https://github.com/udevbe/greenfield/commit/84d63cbe6c73666e832d298abe6dd22175d0efa1

**Reply by Scooter-F** [reply]
It doesn't look to be working for me. I've attached a screenshot in case that's helpful. The red area is the remote application and the black area is my canvas element. 
![greenfield_not_sizing](https://github.com/udevbe/greenfield/assets/36043068/97ed49fe-38f1-46f2-a4fd-1ee01daf98d9)

**Reply by Zubnix** [reply]
I'll pick this one up once https://github.com/udevbe/greenfield/discussions/141 is fixed.

**Reply by Zubnix** [reply]
I've tested the fullscreen scenario for gtk4 & qt application and they behave correctly. The compositor tells the app to go fullscreen and tells the app what resolution the screen is running in. Both gtk & qt then fill the entire screen.

If I understand correctly the issue is that there is a black border for your fullscreen app? I know this happens when applications expect a certain aspect ratio and if the screen does not match, the remaining space is left black. 

More info here: https://wayland.app/protocols/xdg-shell#xdg_toplevel:request:set_fullscreen

> If the surface doesn't cover the whole output, the compositor will position the surface in the center of the output and compensate with with border fill covering the rest of the output. The content of the border fill is undefined, but should be assumed to be in some way that attempts to blend into the surrounding area (e.g. solid black).

There's still some improvement on Greenfield's side though, e.g. centering the content and probably some other stuff.

**Reply by Scooter-F** [reply]
When I start the application windowed, and then maximize it it fills the canvas correctly. When I start the application maximized (which is my actual use case) is how I am getting the black border. 
>  I know this happens when applications expect a certain aspect ratio and if the screen does not match, the remaining space is left black.

When you say screen size here are you referring to the size of the canvas element or the resolution of the monitor I am using at the moment?

**Reply by Scooter-F** [reply]
I have forced my application to expect the same aspect ratio as the canvas I am rendering it into and it still shows the black border.

**Reply by Scooter-F** [reply]
Does the compositor expose any way to maximize an application programmatically? If so that would be an easy fix to my specific issue.

**Reply by Zubnix** [reply]
@Scooter-F There's currently no public api to do that (however it does make sense to add this). However, you can easily modify the compositor to issue a maximize call for each new window by adding 

```
    setTimeout(() => {
      this.setMaximized(true);
    })
```

below this line here:    https://github.com/udevbe/greenfield/blob/master/packages/compositor/src/Desktop.ts#L225

**Reply by Zubnix** [reply]
Happy to investigate your issue further if you can give me a way to reproduce this! (e.g. some code or library I can try).

**Reply by Scooter-F** [reply]
If/when I have the time I'll put together an example project. In the mean time I was able to get where I needed to go by adding the following at the same line you mentioned in your last reply:
```
setTimeout(() => {
      const geometry = this.role.queryGeometry()
      const initialOffset: Point = {x: geometry.position.x * -1, y: geometry.position.y * -1}
      this.role.view.positionOffset = initialOffset
      this.setMaximized(true)
    }, 1000)
```
This  forced my application into the top left corner then maximized it.

**Reply by Scooter-F** [reply]
Alright, I've got the compositor set up so that when the canvas that everything is being rendered in resizes the application is resized and maximized, i.e. always maximized and positioned in the top left corner of the canvas. The problem that I am running into now is that when the canvas is resized to be larger the mouse interactions only work withing the original canvas size. 

So if the canvas starts with dimensions of 100x100 and I resize it to be 200x200 the application I am rendering will be 200x200. However the mouse will only work in the original 100x100 area. @Zubnix Any thoughts?

**Reply by Scooter-F** [reply]
To clarify I can still drag my mouse around, but any listeners I've added, or any clicks do not register when they are out of the original canvas bounds.

**Reply by Zubnix** [reply]
That sounds like a bug 🤔 I'll see if I can reproduce

**Reply by Scooter-F** [reply]
@Zubnix Tried out the full screen mode in the kiosk branch (very cool btw). The error I last mentioned looks to be present there as well. I think my original description of the issue may have been inaccurate though. It looks like the clicks/other events are still firing but the mouse position is not updated beyond the original size of the canvas.

**Reply by Scooter-F** [reply]
Also when the canvas is resized the right edge and bottom edge of the application I'm rendering are once again getting cut off.  Everything looks fine on the initial render.
